### ISSUE TITLE
Is there a way to MPI_Gather with different custom datatypes?

### ISSUE THREAD

**Issue by frizwi** [open-mpi/ompi#12461]
Please submit all the information below so that we can understand the working environment that is the context for your question.

* If you have a problem building or installing Open MPI, [be sure to read this](https://docs.open-mpi.org/en/main/getting-help.html#for-problems-building-or-installing-open-mpi).
* If you have a problem launching MPI or OpenSHMEM applications, [be sure to read this](https://docs.open-mpi.org/en/main/getting-help.html#for-problems-launching-mpi-or-openshmem-applications).
* If you have a problem running MPI or OpenSHMEM applications (i.e., after launching them), [be sure to read this](https://docs.open-mpi.org/en/main/getting-help.html#for-problems-running-mpi-or-openshmem-applications).

### What version of Open MPI are you using? (e.g., v4.1.6, v5.0.1, git branch name and hash, etc.)
OpenMPI 4.1.6

### Describe how Open MPI was installed (e.g., from a source/distribution tarball, from a git clone, from an operating system distribution package, etc.)
Built from source release

### Please describe the system on which you are running

* Operating system/version: Ubuntu 18LTS (I know it's not supported, but don't think my issue is OS related)
* Computer hardware: Single node
* Network type:

-----------------------------

## Details of the problem

Firstly, I want to use MPI_Gather/MPI_Scatter collectives instead of Send/Recv's because I'm under the impression that the collectives will always be performant, and with the later versions there seems to be support for persistent comms which might be even better. If collectives are not necessarily "faster" then, I think I already have a workaround using Send/Recv's so my question is then a bit moot.

What I'm looking for is an MPI_Gather variant that supports a different datatype for each process, I guess it would be called MPI_Gatherw. I have a numerical model with distributed processing, each process sends back a block of different sized data back to the root. This is OK as MPI_Gatherv handles it except that I need to index each block differently, in other words use a different MPI_Type_create_indexed_block for each process's data block. Is there a way to somehow apply a single MPI_Type_index block to the entire recv data at the root? If so, then this might work but as I understand it the MPI_Datatype applies to each block of recv data.

Thanks for any tips!

**Comment by devreal** [comment]
@frizwi The MPI standard explicitly allows different types to be used at senders and the receiver:


> The type signature implied by sendcount, sendtype on MPI process i must be equal
to the type signature implied by recvcounts[i], recvtype at the root. This implies that the
amount of data sent must be equal to the amount of data received, pairwise between each
MPI process and the root. Distinct type maps between sender and receiver are still allowed,
as illustrated in Example 6.6.

My understand is that your use-case is covered by `MPI_GATHERV`. If Open MPI fails to do this correctly then we have to fix it :)

**Comment by frizwi** [comment]
@devreal The statement:

> Distinct type maps between sender and receiver are still allowed, as illustrated
in Example 5.6
from the MPI spec. is a bit misleading, and/or I'm not understanding it properly

The `MPI_Gatherv` definition is as follows:

`int MPI_Gatherv(const void* sendbuf, int sendcount, MPI_Datatype sendtype,
void* recvbuf, const int recvcounts[], const int displs[],
MPI_Datatype recvtype, int root, MPI_Comm comm)`

In other words, a single and same **rectype** is applied to each pairwise transfer, but in my case I need them to be different

**Comment by ggouaillardet** [comment]
There is no `MPI_Gatherw()`-ish collective in the MPI standard that takes an array of `recvtypes` (and `displs` in bytes, similar to `MPI_Alltoallw()`)

**Comment by frizwi** [comment]
@devreal Yep, on the send side there is no issue. It's the `rectype` at the root that I'm having trouble with, `revcounts[i]` with `rectype` needs to match the size of data sent from each process. I'm not sure how to make anything other than a native datatype work at the receiver. Sorry, this is more of an MPI spec. question

**Comment by devreal** [comment]
Sorry, I realize that I misunderstood your question, you had already said that you wanted to use different types at the receiver.

You could receive into a contiguous array and scatter the data from there. I'm not sure why there is no `MPI_Gatherw` that allows you to receive with different types, it might be an interesting proposal to the Forum if there is a good use-case.

**Comment by frizwi** [comment]
Thanks @devreal

I guess now it's a question of performance of Gather plus unpack & distribute data at the root vs `MPI_Send/Recv`, for the latter case I can supply a different derived datatype for each receive.

**Comment by bosilca** [comment]
`MPI_Gatherw` would remove all opportunities for optimizations. With `MPI_Gather` all processes would send the same type signature, each rank knows what the others are sending so they can work as a middleman. With the `MPI_Gatherw` only the root will know what is to be received, so the collective pattern is either a star, which goes back to implementing it with send/recv, or the implementer will have to first do a scatter with the root knowledge (costly).

**Comment by ggouaillardet** [comment]
@bosilca I get your point, but how would `MPI_Gatherw()` be different from `MPI_Gatherv()` performance wise?

**Comment by frizwi** [comment]
> each rank knows what the others are sending so they can work as a middleman.

This is interesting optimisation info, helps me to understand a bit better. So is there presently a performance difference between `Gather` and `Gatherv? If so, I might be better off setting up the gather for the max. worker's case and unpack at the root

**Comment by bosilca** [comment]
It's not as simple as it sounds. The hierarchical gather/gatherv makes sense in a limited number of cases, mostly for hardware with **big** latency and **big** bandwidth (hint __ethernet__), and when the sizes of the contributions are too small to efficiently use the available bandwidth. In such cases, it might be better to coalesce the small messages into a bigger one at the node level (transfer that will use shared memory, so small latency and big bandwidth), and then transfer it over then network as a single, longer message (taking advantage of the bandwidth). 

If the messages of the different processes are few percentage away fro each other, then yes it might make sense to do what you said. Or better, run a benchmark and see how they compare.

**Comment by ggouaillardet** [comment]
Thanks @bosilca, I clearly overlooked the heterogeneous case!

**Comment by frizwi** [comment]
Thanks for all your help, I'll close this now as all my questions have been answered.
